src/gt4py/next/embedded/function_field.py

```diff
@@ -0,0 +1,203 @@
+            raise IndexError("Cannot slice ConstantField without a Domain.")
+        new_domain = embedded_common.sub_domain(self.domain, index)
+        return self.__class__(self.func, new_domain)
+
+    __getitem__ = restrict
+
+    @property
+    def ndarray(self) -> core_defs.NDArrayObject:
+        if _has_empty_domain(self):
+            raise ValueError("Cannot get ndarray for FunctionField without Domain.")
+
+        shape = []
+        for _, rng in self.domain:
+            if Infinity.positive() in (abs(rng.start), abs(rng.stop)):
+                raise ValueError(
+                    f"Cannot construct ndarray with infinite range in Domain: {self.domain}"
+                )
+            else:
+                shape.append(len(rng))
+
+        if self._constant:
+            return np.full(shape, self.func())
+
+        return np.fromfunction(lambda *indices: self.func(*indices), shape)
+
+    def _handle_identity_op(self, other: FunctionField, operator_func: Callable) -> FunctionField:
+        domain_intersection = self.domain & other.domain
+        broadcasted_self = _broadcast(self, domain_intersection.dims)
+        broadcasted_other = _broadcast(other, domain_intersection.dims)
+        return self.__class__(
+            _compose(operator_func, broadcasted_self, broadcasted_other), domain_intersection
+        )
+
+    def _binary_op_wrapper(self, other: FunctionFieldOperand, op: Callable) -> FunctionFieldOperand:
+        if _is_nd_array(other):
+            if _has_empty_domain(self):
+                return self._handle_empty_domain_op(other, op)
+            else:
+                return self._handle_non_empty_domain_op(other, op)
+        elif isinstance(other, self.__class__):
+            return self._handle_identity_op(other, op)
+        else:
+            raise ValueError(
+                f"Unsupported type in binary operation between {self.__class__} and {other.__class__}"
+            )
+
+    def _handle_empty_domain_op(
+        self, other: nd._BaseNdArrayField, op: Callable
+    ) -> nd._BaseNdArrayField:
+        self_broadcasted = self.__class__(self.func, other.domain)
+        new_data = op(self_broadcasted.ndarray, other.ndarray)
+        return other.__class__.from_array(new_data, domain=other.domain)
+
+    def _handle_non_empty_domain_op(
+        self, other: nd._BaseNdArrayField, op: Callable
+    ) -> nd._BaseNdArrayField:
+        domain_intersection = self.domain & other.domain
+        self_broadcasted = self.__class__(self.func, domain_intersection)
+        other_broadcasted = nd._broadcast(other, domain_intersection.dims)
+        other_slices = _get_slices_from_domain_slice(other_broadcasted.domain, domain_intersection)
+        new_data = op(self_broadcasted.ndarray, other_broadcasted.ndarray[other_slices])
+        return other.__class__.from_array(new_data, domain=domain_intersection)
+
+    def _binary_operation(self, op: Callable, other: FunctionFieldOperand) -> FunctionFieldOperand:
```

Comment:
```suggestion
    @overload
    def _binary_operation(self, op: Callable, other: Field) -> Field:
        ...
        
    @overload
    def _binary_operation(self, op: Callable, other: FunctionField) -> FunctionField:
        ...

    def _binary_operation(self, op: Callable, other: Field) -> Field:
```
and remove `FunctionFieldOperand`, the goal is that every Field accepts every other Field (but some might raise because we cannot find a proper operation).